(maint) Improve Kernel.caller() usage

justinstoller opened the pull request:
In Ruby 2, and JRuby 9k, `Kernel#caller` got a new optional argument that
controls how many frames to go back when constructing the call stack.

In most cases where we call `caller` we also throw away most of the
call stack. This has performance implications for every Ruby environment
and especially for JRuby.

Most of the time this doesn't matter because we are getting the call
stack for error reporting, however in at least one case this
information is used for non-fatal warnings.

justinstoller commented on lib/puppet/util/logging.rb:
```diff
@@ -194,9 +194,9 @@ def get_deprecation_offender()
     #  the deprecated method
     if Puppet[:trace]
-      caller()[2..-1]
+      caller(4)
```
Someone should double check my math on this and the one below.

joshcooper replied:
The default signature is `Kernel.caller(start=1, length=nil)`, so previously we skipped the first frame (index 0), and then skipped array entries 0 and 1. So I think we want `caller(3)` which should skip the first 3 frames (0,1,2)?

justinstoller replied:
Thank you, when I rebased I noticed the previous version was incorrect, but everytime I did it in my head I got a different answer.

joshcooper commented on lib/puppet/util/logging.rb:
```diff
@@ -194,9 +194,9 @@ def get_deprecation_offender()
+      caller(4)
     else
-      [caller()[2]]
+      [caller(1, 3)[2]]
```
since we're just returning one frame, I think we can do `[caller(3..3).first]`?

joshcooper commented on lib/puppet/util/log.rb:
```diff
@@ -175,7 +175,7 @@ def Log.coerce_string(str)
     message = _("Received a Log attribute with invalid encoding:%{log_message}") %
         { log_message: Puppet::Util::CharacterEncoding.convert_to_utf_8(str.dump)}
-    message += '\n' + _("Backtrace:\n%{backtrace}") % { backtrace: caller[0..10].join("\n") }
+    message += '\n' + _("Backtrace:\n%{backtrace}") % { backtrace: caller(1, 10).join("\n") }
```
👍 previously there was an off-by-one error (we returned 11 frames instead of 10 as specified in the comments).

joshcooper commented:
Can you also delete the disabled rubocop check?

```
# TODO PUP-8003
Performance/Caller:
  Enabled: false
```

justinstoller commented:
I'll remove the lint, amend the commit with that ticket and look into the block usage it also talks about.

puppetcla commented:
CLA signed by all contributors.

justinstoller commented:
Updated the first commit as discussed. Added one for `sort_by` improvements mentioned in the ticket. I think there's a regression there (potentially w/ the mocks), but my tests are failing locally for other reasons right now, so I'll circle back around to it later.

joshcooper commented:
could you drop the second commit, as I've got a fix for that in PR targeting 5.5.x?

justinstoller commented:
Dropped the second commit and rebased onto HEAD of 6.4.x

joshcooper commented:
Nit the PUP-8003 ticket was moved to the INTL project. Can you just add `(maint)` instead? Otherwise looks great, thanks!

justinstoller commented:
PUP-8003 -> maint
